source/extensions/filters/http/oauth2/filter.cc

```diff
@@ -330,11 +333,25 @@ Http::FilterHeadersStatus OAuth2Filter::decodeHeaders(Http::RequestHeaderMap& he
     const std::string escaped_redirect_uri =
         Http::Utility::PercentEncoding::encode(redirect_uri, ":/=&?");
 
-    const std::string new_url = fmt::format(
-        AuthorizationEndpointFormat, config_->authorizationEndpoint(), config_->clientId(),
-        config_->encodedAuthScopes(), escaped_redirect_uri, escaped_state);
-
-    response_headers->setLocation(new_url + config_->encodedResourceQueryParams());
+    Http::Utility::Url authorization_endpoint;
+    // ASSERT is safe because the constructor throws an exception if the authorization endpoint is
+    // not a valid URL.
+    ASSERT(authorization_endpoint.initialize(config_->authorizationEndpoint(),
+                                             /*is_connect_request=*/false));
+    auto query_params =
+        Http::Utility::parseQueryString(authorization_endpoint.pathAndQueryParams());
```

Comment:
not a huge deal, but we could save this map (`query_params`) in the constructor and just make a copy of the map on every request, should be cheaper than parsing the authorization endpoint on every request. Maybe leave a TODO if you don't think we should handle it now?

Reply:
Done, good feedback. I am also now caching the URL object to avoid parsing the URL on every object, which has the additional benefit of avoiding the ASSERT when using URL parts. Please let me know if you have any concerns.